libcalico-go/lib/backend/k8s/resources/ipam_affinity.go

```diff
@@ -210,6 +251,43 @@ func (c *blockAffinityClient) DeleteKVP(ctx context.Context, kvp *model.KVPair)
 	return c.toV1(kvp)
 }
 
+func (c *blockAffinityClient) deleteKVPV3(ctx context.Context, kvp *model.KVPair) (*model.KVPair, error) {
+	// We need to mark as deleted first, since the Kubernetes API doesn't support
+	// compare-and-delete. This update operation allows us to eliminate races with other clients.
+	var err error
+	nkvp := kvp
+	if kvp.Value == nil {
+		// Need to check if a value is given since V3 deletes are done by key only.
```

Comment:
> 		// Need to check if a value is given since V3 deletes are done by key only.
>		// Look up missing values with the provided key.

What do you mean by this comment, then? It seems to suggest that `kvp.Value` will always be nil in this function. What is the scenario where it isn't going to be nil, if any?

If we are going to do this, we need to ensure that we continue to use the original `kvp.Revision` when calling `Update` below, otherwise we're losing the compare-and-swap behavior that we're guaranteeing to the client.

Reply:
I believe we get to this logic because v3 deletes take in just the name in order to perform their deletes (I think the call that translates that to a KVP is here: https://github.com/projectcalico/calico/blob/master/libcalico-go/lib/clientv3/resources.go#L183). I don't think there should be a case _now_ where they will not be nil, but if we are going to use the `clientv3` calls eventually in our IPAM code (as opposed to the v1 calls we use now), then non-nil values can be passed into this logic.

I agree about the `kvp.Revision` portion. I filled it out without too much thought so I can fix that part up.

Reply:
> I don't think there should be a case now where they will not be nil

We should update the comment to say this then - the code and comment right now suggests that there is a path by which this might be `nil`. 

Either we always query it (with a comment explaining why) or we conditionally query it, but include a comment that says "We expect this to always be nil but we're just being safe"

Reply:
Actually, I think this is all irrelevant anyway - `DeleteKVP` isn't an operation on the v3 client. `v3` is always going to call `Delete`.

Reply:
Which I think is probably enough of a problem to say the IPAM code isn't going to use the clientv3 logic, ever. Adding an additional Get() is just not really worth it, imo.

Reply:
Doesn't the v3 client call `Delete` which then in turn calls `DeleteKVP` here: https://github.com/projectcalico/calico/blob/master/libcalico-go/lib/clientv3/resources.go#L183 ? At least my tests through the v3 client (which makes calls using `Delete`) also ran into these issues because of it.